spyderlib/plugins/shortcuts.py

```diff
@@ -7,179 +7,505 @@
+              padding:0px;
+              border-radius: 0px;
+            }"""
+        self.helper_button.setStyleSheet(style)
+
+        # Layout
+        spacing = 24
+        layout_sequence = QGridLayout()
+        layout_sequence.addWidget(self.label_info, 0, 0, 1, 3)
+        layout_sequence.addItem(QSpacerItem(spacing, spacing), 1, 0, 1, 2)
+        layout_sequence.addWidget(self.label_current_sequence, 2, 0)
+        layout_sequence.addWidget(self.text_current_sequence, 2, 2)
+        layout_sequence.addWidget(self.label_new_sequence, 3, 0)
+        layout_sequence.addWidget(self.helper_button, 3, 1)
+        layout_sequence.addWidget(self.text_new_sequence, 3, 2)
+
+        layout = QVBoxLayout()
+        layout.addLayout(layout_sequence)
+        layout.addSpacing(spacing)
+        layout.addWidget(bbox)
+        self.setLayout(layout)
+
+        # Signals
+        bbox.accepted.connect(self.accept)
+        bbox.rejected.connect(self.reject)
+
+    def keyPressEvent(self, e):
+        """Qt override"""
+        key = e.key()
+
+        self.npressed += 1
+        self.key_non_modifiers.append(key)
+        self.key_modifiers.add(key)
+        self.key_text.append(e.text())
+
+        debug_print('key %s, npressed: %s' % (key, self.npressed))
+
+        if key == Qt.Key_unknown:
+            return
+
+        # The user clicked just and only the special keys
+        # Ctrl, Shift, Alt, Meta.
+        if (key == Qt.Key_Control or
+                key == Qt.Key_Shift or
+                key == Qt.Key_Alt or
+                key == Qt.Key_Meta):
+            return
+
+        # Check if valid keys
+        if key not in VALID_KEYS:
+            return
+
+        modifiers = e.modifiers()
+        if modifiers & Qt.ShiftModifier:
+            key += Qt.SHIFT
+        if modifiers & Qt.ControlModifier:
+            key += Qt.CTRL
+            if sys.platform == 'darwin':
+                self.npressed -= 1
+            debug_print('decrementing')
+        if modifiers & Qt.AltModifier:
+            key += Qt.ALT
+        if modifiers & Qt.MetaModifier:
+            key += Qt.META
+
+        self.keys.add(key)
+
+    def toggle_state(self):
+        """Switch between shortcut entry and Accept/Cancel shortcut mode."""
+        self.edit_state = not self.edit_state
+
+        if not self.edit_state:
+            self.text_new_sequence.setEnabled(False)
+            if self.button_ok.isEnabled():
+                self.button_ok.setFocus()
+            else:
+                self.button_cancel.setFocus()
+        else:
+            self.text_new_sequence.setEnabled(True)
+            self.text_new_sequence.setFocus()
+
+    def nonedit_keyrelease(self, e):
+        """Key release event for non-edit state."""
+        key = e.key()
+
+        if key == Qt.Key_Escape:
+#            self.close()
+            pass
+
+        if key in [Qt.Key_Left, Qt.Key_Right, Qt.Key_Up,
+                   Qt.Key_Down]:
+

```

Comment:
By the way, I don't like this too much. I think just `Up` should take users again to the QLineEdit, not the other keys.